pkg/stanza/fileconsumer/internal/reader/compressedreader.go

```diff
@@ -0,0 +1,147 @@
+package reader
+
+import (
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/header"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/scanner"
+	"go.uber.org/zap"
+	"golang.org/x/net/context"
+	"math"
+	"os"
+)
+
+type CompressedReader struct {
+	*Reader
+
+	buffer           *bytes.Buffer
+	lastReadFileSize int64
+	fromBeginning    bool
+}
+
+func (r *CompressedReader) ReadToEnd(ctx context.Context) {
+	newFileHandle, err := os.Open(r.file.Name())
+	if err != nil {
+		r.set.Logger.Error("problem creating reader for compressed file", zap.Error(err))
+		return
+	}
+	r.file = newFileHandle
+	gzipReader, err := gzip.NewReader(r.file)
+	if err != nil {
+		r.set.Logger.Error("problem creating reader for compressed file", zap.Error(err))
+		return
+	}
+
+	// TODO validate if we can keep this check for the file size to avoid unnecessary decompression
+	// in theory, this check can lead to new log entries going unnoticed, if the new entries are duplicates
+	// of a previous entry, as this leads to the size of the compressed file not being increased.
+	// practically this may not be an issue for logs, since at least the timestamp should change for a new log entry,
+	// even if it has the same message as the line before
+
+	stat, err := r.file.Stat()
+	if err != nil {
+		r.set.Logger.Error("problem creating determining size of compressed file", zap.String("fileName", r.fileName), zap.Error(err))
+		return
+	}
+
+	initalRead := true
+	if lastReadFileSize, ok := r.FileAttributes["lastReadFileSize"]; ok {
+		initalRead = false
+		if lastReadFileSize.(int64) >= stat.Size() {
+			// do not read further unless something has been added to the file
```

Comment:
Similarly, I'm not sure why we necessarily need to use a different strategy for tracking file changes. With non compressed files, we just try to read starting from the offset. If the offset is already at the end, there's nothing to do.

Can we just keep this part the same as well?

Reply:
The main challenge here though is to figure out how to track the current offset within a gzip compressed file, as it does not seem to be possible to start reading from an offset within such a file. However I might be missing something here, so I appreciate any ideas regarding how this could be done.

Reply:
I see your point.

If I'm understanding correctly, gzip can be appended, but I don't know much about the details.

Let's say we have a file which is 100kB. If another 100kB is appended, does it modify the first 100kB in any way, or it is a clean addition? If it's clean, can we read the second half independently starting from a known offset of 100kB? 

If all of the above is correct, then I can see how tracking the size of the file is effectively tracking the offset. In that case, I think we could find a way to put offset tracking behind an interface, so that when the user selects gzip, we not only change the type of reader, but also the mechanism for offset tracking. Ideally these both fit into a shared framework though.

If any of my assumptions are incorrect, or if this just seems too complicated, we could consider initially just supporting gzip with the caveat that appended data is ignored.

Reply:
I did some more research about gzip and it seems you can also append to an existing gzip file, without altering its previous file content. I will try to use an `io.SectionReader` within the `gzip.Reader` to be able to read from the offset of the previous iteration - in theory that should hopefully yield the desired behavior and enable us to change as little as possible. If it should turn out too complicated I think we could - at least for a first version - also go for a solution where appended data is ignored, if the arguments in the last paragraph of my comment [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33406#issuecomment-2154051222) make sense

Reply:
alright, seems like using the `io.SectionReader` together with `gzip.Reader` does the trick, if something got appended to the observed gzip file, so this way we are able to pick up at the last offset. One limitation of course is if the whole original content got recompressed - but i think that will not be a realistic scenario to happen.

I will do the proper implementation using this approach now and give you an update as soon as the PR is ready for another review :)